### PR TITLE
Add Set IpSet and Envelop data types

### PR DESCRIPTION
**Pull request by davidhadas**
Add 3 more data types that depend only on Count that was already merged.
- Set that can take a single token or a slice of tokens
- IpSet that supports IPv4 and IPv6 and can take IP as a string, net.IP or a slice of net.IPs
- Envelop that represents session data (req/response) beyond what we have in the actual request or the actual body. Right now Envelop includes only timing of the session, but this may grow.

This PR can be reviewed and merged independently from [PR/16](https://github.com/knative-sandbox/security-guard/pull/16)

### CODE REVIEW

**Review by psschwei — COMMENTED** [review]
couple of nits

File: pkg/apis/wsecurity/v1alpha1/ipSet.go [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+		}
+	}
+}
+
+//////////////////// CidrSetConfig ////////////////
+
+type CIDR net.IPNet
+
+// Return true if cidr include the ip range of otherCidr
+func (cidr *CIDR) Include(otherCidr CIDR) bool {
```
[the remark this answers is not in the document]
Comment (davidhadas) [reply pkg/apis/wsecurity/v1alpha1/ipSet.go:108]: The current way aligns better with the way we pile.Merge() and config.Fuse() - I don't see a reason to do an exception here

File: pkg/apis/wsecurity/v1alpha1/ipSet.go [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+// The maximal extension allowed by the implementation is a C Subnet
+// (i.e. mask of 255.255.255.0 in IPv4)
+// InflateBy returns true if successful
+func (cidr *CIDR) InflateBy(ip net.IP) bool {
+	if (*net.IPNet)(cidr).Contains(ip) {
+		return true
+	}
+
+	cidrLen := len(cidr.IP)
+	if len(ip) != cidrLen {
```
[the remark this answers is not in the document]
Comment (davidhadas) [reply pkg/apis/wsecurity/v1alpha1/ipSet.go:135]: If the method does not reach line 153, it means it has returned false and  cidr was not changed.
If the method reach line 153, it means it has returned true and cidr was changed (inflated).
```golang
        // Inflate the CIDR to cover the IP as well!
	xor := cidr.IP[cidrLast] ^ ip[cidrLast]
	bitsShared := bits.LeadingZeros8(xor)
	mask := net.CIDRMask(cidrBits-8+bitsShared, cidrBits)
	cidr.IP = cidr.IP.Mask(mask)
	cidr.Mask[cidrLast] &= mask[cidrLast]
	return true
```
So apperenty we are fine.

**Comment by davidhadas** [comment]
Aligned with PR/20

**Review by psschwei — COMMENTED** [review]
/lgtm

**Pull request merged**